### PR TITLE
Email Notification for volunteers

### LINKED ISSUE #714: Email notification to volunteers
## Description
As a USER,
I need to receive email notifications for shift assignment and it's cancellation,
so that I can update my scheduler with the same.

### Update [Required]
- [x] [Email notification for shift assignement]
- [x] [Email notification for shift cancellation]

## Definition of Done
- [x] All of the required items are completed.
- [x] Approval by 2 mentors.

## Estimation
3 hours

### PR DESCRIPTION
**Pull request by anjali-dhanuka**
# Description
The volunteers are notified by an email whenever a shift has been assigned to them or it is cancelled.

Fixes #714 

# Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

# How Has This Been Tested?
image

# Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged 

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by anjali-dhanuka** [comment]
Hi @abhishekarya286 @ayushgarg1804 @kamsuri Please review. Also please answer this https://systers-opensource.slack.com/archives/C0FSZS007/p1528315423000785

File: vms/shift/views.py [hunk 1]
```diff
@@ -192,6 +194,9 @@ def cancel(request, shift_id, volunteer_id):
             try:
                 cancel_shift_registration(volunteer_id, shift_id)
                 if admin:
+                    vol_email = Volunteer.objects.get(pk=volunteer_id).email
+                    send_mail("Shift Cancelled", "Your shift has been cancelled",
+                              "messanger@localhost.com", [vol_email], fail_silently=False)
```
Comment (abhishekarya286) [inline vms/shift/views.py]: Change it to `email_host'. And, how would you authenticate password at SMTP servers?

File: vms/shift/views.py [hunk 2]
```diff
@@ -611,6 +616,10 @@ def sign_up(request, shift_id, volunteer_id):
                     result = register(volunteer_id, shift_id)
                     if result == "IS_VALID":
                         if admin:
+                            vol_email = Volunteer.objects.get(pk=volunteer_id).email
+                            send_mail("Shift Assigned", "A shift has been assigned to you.",
+                                       "messanger@localhost.com", [vol_email],
```
Comment (abhishekarya286) [inline vms/shift/views.py]: Ditto

File: vms/shift/views.py (hunk 1 above)
Comment (anjali-dhanuka) [reply vms/shift/views.py]: what do you mean by changing to email_host?

Comment (anjali-dhanuka) [reply vms/shift/views.py]: We can authenticate passwords by encrypting it and return the original password to a variable. https://stackoverflow.com/questions/42030080/encypting-smtp-server-password-in-django-settings

**Comment by anjali-dhanuka** [comment]
@ayushgarg1804 @abhishekarya286 @kamsuri I have changed the messages in emails. Please review.

File: vms/shift/views.py (hunk 1 above)
Comment (abhishekarya286) [reply vms/shift/views.py]: I was talking about the `email_host` variable in settings.py file. In this way, the admin will be able to select the email id dynamically. What's your say on this?

File: vms/shift/templates/shift/cancel_email.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{% autoescape off %}
+{{admin_first_name}} {{admin_last_name}} has cancelled your shift starting from {{shift_start_time}} to {{shift_end_time}} of {{job_name}} under {{event_name}} event. For any questions or concerns, email {{admin_email}}.
```
Comment (abhishekarya286) [inline vms/shift/templates/shift/cancel_email.txt]: shouldn't it  be `during the {{event_name}} event`?

Comment (abhishekarya286) [reply vms/shift/templates/shift/cancel_email.txt]: You should add date too.

Comment (anjali-dhanuka) [reply vms/shift/templates/shift/cancel_email.txt]: done @abhishekarya286 Please review

File: vms/shift/views.py (hunk 2 above)
Comment (anjali-dhanuka) [reply vms/shift/views.py]: done

**Comment by anjali-dhanuka** [comment]
@abhishekarya286 Done with the changes you suggested, @ayushgarg1804 @kamsuri Please review.

**Comment by abhishekarya286** [comment]
**LGTM**. Thanks @anjali-dhanuka. @kamsuri @ayushgarg1804. **PTAL**.

File: vms/shift/views.py [hunk 4]
```diff
@@ -611,6 +632,26 @@ def sign_up(request, shift_id, volunteer_id):
...
+                                    'shift_start_time': shift_object.start_time,
+                                    'shift_end_time': shift_object.end_time,
+                                    'admin_email': admin.email,
+                                    'job_name': job_object.name,
+                                    'event_name': event_object.name,
+                                    'shift_date': shift_object.date,
+                                    })
+                            send_mail("Shift Assigned", message,
+                                       "messanger@localhost.com", [vol_email],
+                                       fail_silently=False)
```
Comment (ayushgarg1804) [inline vms/shift/views.py]: should catch exception

File: vms/shift/views.py [hunk 5]
```diff
@@ -14,6 +14,8 @@
 from django.views.generic import TemplateView, DeleteView, ListView
 from django.views.generic.edit import FormView, UpdateView
 from django.utils.decorators import method_decorator
+from django.core.mail import send_mail
+from django.template.loader import render_to_string
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/shift/views.py]: done @ayushgarg1804

File: vms/shift/views.py (hunk 4 above)
Comment (anjali-dhanuka) [reply vms/shift/views.py]: done @ayushgarg1804

**Comment by anjali-dhanuka** [comment]
@kamsuri @ayushgarg1804 Done with the changes you requested. Please review.

**Comment by anjali-dhanuka** [comment]
@abhishekarya286 @ayushgarg1804 @kamsuri I have added check on the no of emails sent, the email recipient. Should we add the check for the email content too?

**Comment by anjali-dhanuka** [comment]
@ayushgarg1804 @abhishekarya286 @kamsuri Done with adding tests, please review.

File: vms/shift/tests/test_manageVolunteerShift.py [hunk 6]
```diff
@@ -374,11 +412,30 @@ def test_cancel_assigned_shift(self):
...
+        mail.outbox = []
+        mail.send_mail("Shift Cancelled", "message",
+                              "messanger@localhost.com", [volunteer_1.email])
+        self.assertEqual(len(mail.outbox), 1)
+        msg = mail.outbox[0]
+        self.assertEqual(msg.subject, 'Shift Cancelled')
+        self.assertEqual(msg.to, ['volunteer-email@systers.org'])
+        '''url = 'https://example.com%s' % reverse('url_name')
+        self.assertIn(url, msg.body)  # verification urls, change password urls etc. when applicable
+        self.assertIn('Specific text like comments on why something was accepted/rejected', msg.body)'''
```
[the remark this answers is not in the document]
Comment (ayushgarg1804) [reply vms/shift/tests/test_manageVolunteerShift.py]: @Monal5031

[the remark this answers is not in the document]
Comment (Monal5031) [reply vms/shift/tests/test_manageVolunteerShift.py]: These are anjali's changes @ayushgarg1804 :)

**Comment by abhishekarya286** [comment]
Hi @anjali-dhanuka. Can you squash your commits? Try to have a maximum of 2. Reviewing PR is really difficult in case of more than two commits.

File: vms/shift/templates/shift/cancel_email.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{% autoescape off %}
+{{admin_first_name}} {{admin_last_name}} has cancelled your shift starting from {{shift_start_time}} to {{shift_end_time}} on {{shift_date}} of {{job_name}} during the {{event_name}} event. For any questions or concerns, email {{admin_email}}.
```
Comment (kamsuri) [inline vms/shift/templates/shift/cancel_email.txt]: {{admin_first_name}} {{admin_last_name}} has cancelled your shift scheduled for {{shift_date}} from {{shift_start_time}} to {{shift_end_time}} of {{job_name}} for the event {{event_name}}. For any questions or concerns, email {{admin_email}}.
I think you can change the message to above sentence.

File: vms/shift/templates/shift/sign_up_email.txt [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{% autoescape off %}
+{{admin_first_name}} {{admin_last_name}} has assigned you a shift starting from {{shift_start_time}} to {{shift_end_time}} on {{shift_date}} of {{job_name}} during the {{event_name}} event. For any questions or concerns, email {{admin_email}}.
```
Comment (kamsuri) [inline vms/shift/templates/shift/sign_up_email.txt]: has assigned you a shift scheduled for ..
Same as above.

**Comment by anjali-dhanuka** [comment]
@kamsuri Done with the changes. @abhishekarya286 @ayushgarg1804 PLease review so that we can merge this.

**Comment by kamsuri** [comment]
@ayushgarg1804 @abhishekarya286 we need to rebase and merge, right?

File: vms/shift/tests/test_manageVolunteerShift.py [hunk 9]
```diff
@@ -467,6 +476,19 @@ def test_cancel_assigned_shift(self):
...
```
[the remark this answers is not in the document]
Comment (nida) [reply vms/shift/tests/test_manageVolunteerShift.py]: @anjali-dhanuka Can you address this issue?

**Pull request merged**